File: Practice4/lab4.m
function main
    sequence_length = 31;

    % Первая последовательность Голда (x = 8, y = 15)
    x1 = [0 1 0 0 0]; % 01000
    y1 = [0 1 1 1 1]; % 01111
    gold_sequence1 = generateGoldSequence(x1, y1, sequence_length);

    % Вторая последовательность Голда (x = 9, y = 10)
    x2 = [0 1 0 0 1]; % 01001
    y2 = [0 1 0 1 0]; % 01010
    gold_sequence2 = generateGoldSequence(x2, y2, sequence_length);

    % Вывод первой и второй последовательности
    disp('Первая последовательность Голда (x=8, y=15):');
    fprintf('%d', gold_sequence1);
    fprintf('\n');

    disp('Вторая последовательность Голда (x=9, y=10):');
    fprintf('%d', gold_sequence2);
    fprintf('\n');

    % Автокорреляция с проверкой на m-последовательность
    disp('Автокорреляция первой последовательности с проверкой на m-последовательность:');
    calculateAutocorrelationAndCheckMSequence(gold_sequence1, sequence_length);

    % Взаимная корреляция между первой и второй последовательностями
    cross_correlation = correlation(gold_sequence1, gold_sequence2, sequence_length);
    fprintf(['\nВзаимная корреляция между первой и второй последовательностями: ', num2str(cross_correlation), '/', num2str(sequence_length)]);
    fprintf('\n');

    % Графическая визуализация
    visualizeAutocorrelation(gold_sequence1);
    visualizeCrossCorrelation(gold_sequence1, gold_sequence2);
end

% Функция для генерации последовательности Голда
function gold_sequence = generateGoldSequence(x, y, length)
    gold_sequence = zeros(1, length);
    for i = 1:length
        new_bit_x = xor(x(5), x(3)); % XOR элементов 5 и 3
        new_bit_y = xor(y(5), y(3)); % XOR элементов 5 и 3
        gold_sequence(i) = xor(x(5), y(5)); % Голд последовательность как XOR выходов двух регистров

        % Обновляем регистры
        x = [new_bit_x, x(1:4)];
        y = [new_bit_y, y(1:4)];
    end
end

% Проверка сбалансированности
function balanced = checkBalance(sequence)
    count_ones = sum(sequence == 1);
    count_zeros = length(sequence) - count_ones;
    balance = abs(count_ones - count_zeros);
    if balance <= 1
        balanced = true;
    else
        balanced = false;
    end

    if balance
       fprintf('\nm-последовательность');
    else
       fprintf('\nне m-последовательность');
    end
end

% Проверка на m-последовательность
function is_m = isMSequence(sequence)
    if length(sequence) ~= 31
        is_m = false;
    else
        is_m = checkBalance(sequence);
    end
end

% Функция для автокорреляции с проверкой на m-последовательность
function calculateAutocorrelationAndCheckMSequence(sequence, sequence_length)
    for shift = 0:sequence_length - 1
        shifted_sequence = circshift(sequence, [0 shift]);
        corr = calculateCorrelation(sequence, shifted_sequence);
        is_m_sequence = isMSequence(shifted_sequence);

        fprintf('%4d | %s | %4d/%d | %s\n', shift, sprintf('%d ', shifted_sequence), corr, is_m_sequence);
    end
end

% Расчет корреляции между двумя последовательностями
function corr = correlation(seq1, seq2, length)
    corr = sum(seq1(1:length) .* seq2(1:length));
end

% Расчет корреляции между двумя последовательностями
function corr = calculateCorrelation(seq1, seq2)
    corr = sum(seq1 == seq2) - sum(seq1 ~= seq2);
end

% Визуализация автокорреляции
function visualizeAutocorrelation(sequence)
    sequence_length = length(sequence);
    autocorr_values = zeros(1, sequence_length);

    for shift = 0:sequence_length
        shifted_sequence = circshift(sequence, [0 shift]);
        autocorr_values(shift + 1) = calculateCorrelation(sequence, shifted_sequence);
    end

    figure;
    stem(0:sequence_length, autocorr_values, 'filled', 'red');
    title('Autocorrelation of Gold Sequence');
    xlabel('Shift');
    ylabel('Correlation');
    grid on;
end

% Визуализация взаимной корреляции
function visualizeCrossCorrelation(seq1, seq2)
    sequence_length = length(seq1);
    crosscorr_values = zeros(1, sequence_length);

    for shift = 0:sequence_length
        shifted_seq2 = circshift(seq2, [0 shift]);
        crosscorr_values(shift + 1) = correlation(seq1, shifted_seq2, sequence_length);
    end

    figure;
    stem(0:sequence_length, crosscorr_values, 'filled', 'green');
    title('Cross-Correlation of Two Gold Sequences');
    xlabel('Shift');
    ylabel('Correlation');
    grid on;
end

File: Practice4/lab4.cpp
#include <iostream>
#include <vector>
#include <iomanip>
#include <cmath>

// Функция для генерации последовательности Голда
std::vector<int> generateGoldSequence(std::vector<int> x, std::vector<int> y, int length) {
    std::vector<int> gold_sequence;
    for (int i = 0; i < length; ++i) {
        int new_bit_x = x[4] ^ x[2]; // XOR элементов 5 и 3
        int new_bit_y = y[4] ^ y[2]; // XOR элементов 5 и 3
        gold_sequence.push_back(x[4] ^ y[4]); // Голд последовательность как XOR выходов двух регистров

        // Обновляем регистры
        for (int j = 4; j > 0; --j) {
            x[j] = x[j - 1];
            y[j] = y[j - 1];
        }
        x[0] = new_bit_x;
        y[0] = new_bit_y;
    }
    return gold_sequence;
}

// Функция для циклического сдвига вектора на одну позицию вправо
std::vector<int> cyclicShift(const std::vector<int>& sequence) {
    std::vector<int> shifted_sequence = sequence;
    int last_bit = shifted_sequence.back();
    shifted_sequence.pop_back();
    shifted_sequence.insert(shifted_sequence.begin(), last_bit);
    return shifted_sequence;
}

// Функция для расчёта корреляции между двумя последовательностями
int calculateCorrelation(const std::vector<int>& seq1, const std::vector<int>& seq2) {
    int correlation = 0;
    for (size_t i = 0; i < seq1.size(); ++i) {
        correlation += (seq1[i] == seq2[i]) ? 1 : -1;
    }
    return correlation;
}

// Сбалансированность
bool checkBalance(const std::vector<int>& sequence) {
    int count_ones = 0;
    for (int bit : sequence) {
        if (bit == 1) count_ones++;
    }
    int count_zeros = sequence.size() - count_ones;
    return std::abs(count_ones - count_zeros) <= 1;
}

// Проверка на m-последовательность
bool isMSequence(const std::vector<int>& sequence) {
    if (sequence.size() != 31) return false; // Проверка длины последовательности 2^n - 1
    return checkBalance(sequence); // Дополнительные критерии для m-последовательности
}

// Вычисление автокорреляции с проверкой на m-последовательность
void calculateAutocorrelationAndCheckMSequence(const std::vector<int>& sequence, int sequence_length) {
    std::vector<int> shifted_sequence = sequence;
    std::cout << "Сдвиг|                                                               |Корреляция| m-последовательность\n";
    for (int shift = 0; shift < sequence_length; ++shift) {
        int corr = calculateCorrelation(sequence, shifted_sequence);
        
        std::cout << std::setw(4) << shift << " | ";
        for (int bit : shifted_sequence) {
            std::cout << bit << " ";
        }
        std::cout << "| " << corr << "/" << sequence_length;
        
        // Проверка на m-последовательность
        bool is_m_sequence = isMSequence(shifted_sequence);
        std::cout << "    | " << (is_m_sequence ? "Да" : "Нет") << "\n";
        
        shifted_sequence = cyclicShift(shifted_sequence);
    }
}


   return 0;
}


